Add Travertino to Toga repository

HalfWhitt opened the pull request:
Fixes https://github.com/beeware/travertino/issues/246

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

freakboy3742 commented on core/tests/travertino/__init__.py:
In the interests of ensuring a clear separation between packages, it feels to me like the travertino tests should be part of the travertino package, and the CI matrix that currently runs the core tests should be extended to also do a travertino test while they're at it.

That likely means duplicating the coverage pieces as well, so there's a coverage report of Travertino independent of Core - although that likely won't pass at present because we don't have 100% branch coverage.

HalfWhitt replied:
Hm, okay. I'll have to figure out how to do that.

How will that affect command-line arguments to tox/pytest? Will `-- tests/foo` match in both Toga *and* Travertino? I mean, I guess there's one way to find out...

freakboy3742 replied:
You'll likely need an additional set of tox targets to handle Travertino, due to the different working directory and coverage collection.

HalfWhitt replied:
Or, I guess I'm conflating CI and manual invocation of tox...

mhsmith replied:
Travertino contains no platform-specific code, so it shouldn't need to be tested on every platform, only on every Python version.

In fact, I would question whether even toga-core needs to be tested on every platform. There are only a few platform-dependent tests around `get_current_platform`, and it should be possible for those to achieve 100% coverage by simulating the platform with monkey patches, as they already do for Android, iOS and web. This would have the advantage of allowing the code for all platforms to be tested on any platform.

HalfWhitt replied:
> Travertino contains no platform-specific code, so it shouldn't need to be tested on every platform, only on every Python version.

That's a fair point. I didn't think about the fact that there wasn't a reason to extend the current Toga-testing pattern to it.

> In fact, I would question whether even toga-core needs to be tested on every platform. There are only a few platform-dependent tests around `get_current_platform`, and it should be possible for those to achieve 100% coverage by simulating the platform with monkey patches, as they already do for Android, iOS and web. This would have the advantage of allowing the code for all platforms to be tested on any platform.

How do you feel about this, @freakboy3742 ?

freakboy3742 replied:
There's enough platform specific stuff in toga core (e.g., path handling) that I'd feel better retaining a spread of platform tests, rather than mocking the platform parts. However, Travertino is "pure math" - I can't think of anything that is platform specific, so it only needs a single test AFAICT.

mhsmith replied:
OK, makes sense.

HalfWhitt commented:
@freakboy3742 I'm not sure if you're waiting on this to do the new Toga release by any specific time, but in case you are, heads up that I'll have little to no free time for it in the next week or so.

freakboy3742 commented:
@HalfWhitt Thanks for the heads up. If the need for a 0.5 release becomes more pressing, we'll build on the work you've done here.

HalfWhitt commented:
@freakboy3742 Any idea what might be causing the Linux testbed failures? They consistently hang up right after tests/test_command.py::test_unknown_system_command, so presumably they're somehow getting stuck on tests/test_font.py::test_use_system_font_fallback. I don't see an existing bug filed for that, and I can't make out how anything I've done would affect it.

freakboy3742 commented:
> @freakboy3742 Any idea what might be causing the Linux testbed failures? They consistently hang up right after tests/test_command.py::test_unknown_system_command, so presumably they're somehow getting stuck on tests/test_font.py::test_use_system_font_fallback. I don't see an existing bug filed for that, and I can't make out how anything I've done would affect it.

I can't say I've seen that one before. If I had to guess, the exception at the start of the test run has something to do with it:
```
Traceback (most recent call last):
  File "/home/runner/work/toga/toga/testbed/build/testbed/ubuntu/noble/testbed-0.0.1/usr/lib/testbed/app_packages/toga_gtk/container.py", line 145, in do_get_preferred_height
    self.recompute()
  File "/home/runner/work/toga/toga/testbed/build/testbed/ubuntu/noble/testbed-0.0.1/usr/lib/testbed/app_packages/toga_gtk/container.py", line 106, in recompute
    self._content.interface.style.layout(self._content.interface, self)
TypeError: Pack.layout() takes 2 positional arguments but 3 were given
Traceback (most recent call last):
  File "/home/runner/work/toga/toga/testbed/build/testbed/ubuntu/noble/testbed-0.0.1/usr/lib/testbed/app_packages/toga_gtk/container.py", line 176, in do_size_allocate
    self._content.interface.style.layout(self._content.interface, self)
TypeError: Pack.layout() takes 2 positional arguments but 3 were given
--ci: Don't know how to open documents with extension 
Waiting for app to be ready for testing... Traceback (most recent call last):
  File "/home/runner/work/toga/toga/testbed/build/testbed/ubuntu/noble/testbed-0.0.1/usr/lib/testbed/app_packages/toga_gtk/container.py", line 176, in do_size_allocate
    self._content.interface.style.layout(self._content.interface, self)
TypeError: Pack.layout() takes 2 positional arguments but 3 were given
```

That's occurring before the test suite starts, which suggests the actual test app may not be successfully instantiating. 

The test suite can sometimes get into an odd state if it's unable to create a window *at all* - and the test that is locking looks like it could be the first test that actually requires a widget to exist. If something in the test setup is throwing an exception that is being swallowed because of all the layers of async, the test might be awaiting a widget (or window) that never gets created.

The other thing pointing at this as a possible cause is the failure of the `tests/app/test_desktop.py::test_presentation_mode` test. It might be that this test is failing in a way that causes the next "window using" test to fail.

HalfWhitt commented:
> I can't say I've seen that one before. If I had to guess, the exception at the start of the test run has something to do with it:

That was it, thank you!

HalfWhitt commented:
Alright, this is at least *working*. I'm happy for any feedback, particularly on how I've organized the tox environments, and whether it might be worth collapsing the core tests down to one platform like @mhsmith mentioned [above](https://github.com/beeware/toga/pull/3086#discussion_r1921615918).

mhsmith commented on .github/workflows/ci.yml:
```diff
@@ -77,8 +78,18 @@ jobs:
         platform: [ "macos-latest", "ubuntu-latest", "windows-latest" ]
         python-version: [ "3.9", "3.10", "3.11", "3.12", "3.13" ]
+        package: ["core", "travertino"]
+        exclude:
+        - package: travertino
+          platform: macos-latest
+        - package: travertino
+          platform: ubuntu-latest
```
When we run a job on only one platform, we usually use Ubuntu because it has the least availability issues.

HalfWhitt replied:
Ah, noted. I only picked Windows since its minutes are cheaper, didn't realize they vary in availability.

mhsmith commented on core/src/toga/style/pack.py:
```diff
@@ -296,19 +342,7 @@ def apply(self, prop: str, value: object) -> None:
                 self._applicator.refresh()
 
-    def layout(self, viewport: Any, _deprecated_usage=None) -> None:
-        ######################################################################
-        # 2024-12: Backwards compatibility for Travertino 0.3.0
-        ######################################################################
```
There are a couple of other Travertino 0.3 compatibility blocks in this file – can they be removed as well?

* START/END at line 67
* "accessing directional properties via bracket notation" at line 271

HalfWhitt replied:
Ah, yes, thank you. I missed those.

HalfWhitt commented:
I've started writing the remaining tests necessary to get Travertino to 100% coverage. I'm not sure if it would make more sense to include them here or as a separate PR afterward.

mhsmith commented:
This PR is pretty big already, so I would stack a separate PR on top of it which reaches 100% coverage and enables coverage enforcement in the CI.

There are also a few conflicts now from #3113, but I'll try to avoid causing any more until after this PR is merged.

freakboy3742 commented on tox.ini:
```diff
@@ -9,13 +9,13 @@ extend-ignore =
 [tox]
 envlist = py{39,310,311,312,313}-cov,coverage
```
```suggestion
envlist = py{39,310,311,312,313}-cov{,-trav},coverage
```

freakboy3742 commented on tox.ini:
```diff
@@ -90,6 +100,7 @@ suicide_timeout = 1
 deps =
     # editable install so docstrings can be updated for 'all' and 'live'
     -e {tox_root}{/}core[docs]
+    {tox_root}{/}travertino
```
Should this not be an editable install as well?

HalfWhitt replied:
So it should!

freakboy3742 commented on travertino/pyproject.toml:
```diff
@@ -0,0 +1,90 @@
+]
+
+[project.optional-dependencies]
+# Extras used by developers *of* Toga are pinned to specific versions to
```
Not sure what has happened here - these look like the same dependencies (and docstring) for Toga core. It makes sense why typing-extensions and coverage is needed; but why pytest-asyncio and pytest-freezer?

freakboy3742 commented on travertino/src/travertino/__init__.py:
```diff
@@ -0,0 +1,17 @@
+
+    # Excluded from coverage because a pure test environment (such as the one
+    # used by tox in CI) won't have setuptools_scm
+    __version__ = get_version("../..", relative_to=__file__)  # pragma: no cover
```
```suggestion
    __version__ = get_version("../../..", relative_to=__file__)  # pragma: no cover
```

freakboy3742 reviewed (changes requested):
This all looks pretty good - nice work! (and hoo-boy does Github not like the size of this PR... 😅)

I've flagged a few minor things inline. The only minor thing I couldn't comment on inline (because there's no changes to the file that needs the comment) - the dependabot configuration (`.github/dependabot.yaml`) needs an update to include travertino.

The only significant issue I spotted was about towncrier/change notes:
- There's a bunch of change notes in the current Travertino repo; those haven't made the transition over, so all the features added since Travertino 0.3.0 won't be reflected anywhere when Travertino 0.5.0 is released.
- The *historical* list of change notes has also been lost, so if you wanted to know what was added in Travertino 0.3.0, there's no way to find out.
- Going forward, it's not clear where *new* change notes would fit. CI will require a changenote, but they'll currently be incorporated into the Toga change notes... which isn't *wrong*, but isn't really helpful either.

My suggestion would be:
1. Restore/merge over the CHANGELOG.rst file from the current Travertino repository.
2. Add new "classes" of Towncrier message for `trav-bugfix`, `trav-feature` and `trav-removal`. (No need for docs, as there's no Travertino docs; and no need for misc as they're housekeeping).
3. Migrate all the current change notes into toga's changes notes as their equivalent `trav-` tags, referencing the current Travertino repo ticket numbers.

When I cut a release, I run towncrier to generate the release notes; I can manually copy-and-paste the Travertino issues (and update links) into the travertino CHANGELOG file. The 0.5.0 release will need some special handling because all the links will need to point to the Travertino repo; however, once that's archived, the tickets will all be Toga tickets (and we should make a note to migrate all the current tickets over to Toga's repo before archiving Travertino).

One other update (or pair of updates) that is required - we need to update the release guide (`docs/howto/internals/release`) to cover the extra travertino steps, and the contribution guide to cover how to test Travertino. I'd suggest a "contributing to Travertino" section that mostly defers to the code contribution docs, but notes the installation differences, tox entry points, and changenote suffixes.

mhsmith commented:
> The only significant issue I spotted was about towncrier/change notes

Does Travertino need a separate changelog at all in the future? If a Travertino change affects Toga's public API (which is the case for most items in Travertino's existing changelog), then it should go into the Toga changelog. If it doesn't affect Toga's public API, then it doesn't need a public announcement any more than a change to Toga's backend API would.

HalfWhitt commented:
> > The only significant issue I spotted was about towncrier/change notes
> 
> Does Travertino need a separate changelog at all in the future? If a Travertino change affects Toga's public API (which is the case for most items in Travertino's existing changelog), then it should go into the Toga changelog. If it doesn't affect Toga's public API, then it doesn't need a public announcement any more than a change to Toga's backend API would.

This was how I was envisioning it going forward, though I forgot I'd still need to include all of Travertino's outstanding post-0.3.0 notes. Rather than writing agnostically for theoretical Travertino users at large, I think it makes more sense (at least for now) to target release notes based on how, and if, changes ultimately affect Toga users. For instance, the bugfix https://github.com/beeware/travertino/pull/224 is much more relevant in Toga's release notes.

This would mean, of course, that I'd have to evaluate each note on a case by case basis, categorizing as appropriate, when copying them over. And I'm not sure how, if at all, to handle Travertino's existing prior change log in this scheme. Perhaps keep it in Travertino, with a note at the top about it being archived from its prior separate existence?

HalfWhitt commented on .github/workflows/ci.yml:
```diff
@@ -77,8 +78,18 @@ jobs:
         include:
         - experimental: false
+        - package: "core"
+          travertino: ""
+        - package: "travertino"
+          travertino: "-trav"
```
This was my first thought, guess I should've stuck with it.

mhsmith commented:
> This would mean, of course, that I'd have to evaluate each note on a case by case basis, categorizing as appropriate, when copying them over.

Ignoring the `misc` items, which are mostly dependabot updates, there are only 15.

> And I'm not sure how, if at all, to handle Travertino's existing prior change log in this scheme. Perhaps keep it in Travertino, with a note at the top about it being archived from its prior separate existence?

Seems reasonable.

freakboy3742 commented:
> > Does Travertino need a separate changelog at all in the future? If a Travertino change affects Toga's public API (which is the case for most items in Travertino's existing changelog), then it should go into the Toga changelog. If it doesn't affect Toga's public API, then it doesn't need a public announcement any more than a change to Toga's backend API would.
...
> ... I think it makes more sense (at least for now) to target release notes based on how, and if, changes ultimately affect Toga users. For instance, the bugfix [beeware/travertino#224](https://github.com/beeware/travertino/pull/224) is much more relevant in Toga's release notes.

That's fair. It's also going to look odd for most of the Travertino release notes to say "0.5.3 - no changes", as most Toga releases likely won't have Travertino fixes at all (based on historical cadence).

If/when we ever extract Travertino into a standalone repo, the gap in the release notes will be odd; but there won't be any real trigger to do this until Colosseum (or similar) is viable - which also means there won't be any real version cross-project compatibility concerns until this happens, removing the real value of having an independent set of change notes.

So - lets go with:
1. Run towncrier --draft in the Travertino repo to generate 0.5.0 updates to the travertino CHANGELOG
2. Migrate the updated CHANGELOG file to the toga/travertino repo
3. Add a note to the top of the migrated file that notes the repo migration and new release policy, directing to the Toga release notes page.

HalfWhitt commented:
> So - lets go with:
> 
> 1. Run towncrier --draft in the Travertino repo to generate 0.5.0 updates to the travertino CHANGELOG
> 2. Migrate the updated CHANGELOG file to the toga/travertino repo
> 3. Add a note to the top of the migrated file that notes the repo migration and new release policy, directing to the Toga release notes page.

Is that instead of or in addition to adding those changes (or at least ones relevant to Toga users) in Toga's 0.5.0 changes?

...Now that I think about, I have no idea if it's even possible to tell Towncrier to link notes to PRs in another repository. Maybe that's trouble than it's worth.

freakboy3742 commented:
> Is that instead of or in addition to adding those changes (or at least ones relevant to Toga users) in Toga's 0.5.0 changes?

No - I'm thinking of the documentation of Travertino 0.5.0 as if it were generated from the Travertino repo. The features that have been migrated in this PR were all developed "over there". Going forward, we only have Toga changenotes for Travertino features.

HalfWhitt commented:
Okay, I *think* I've addressed everything:
- Add Travertino to the Dependabot config
- Remove unnecessary test dependencies from Travertino
- Add Travertino's changelog, including 0.5.0 and a note explaining the move
  (It has today's date for 0.5.0; we should remember to change it when pushing the actual release.)
- Add Travertino to the How-To section
  Installing a Toga environment installs Travertino, and we're using the same changelog system, so it doesn't really need any separate mention for how to install and make changes; I've added info about how to run its test suite. `-m test` now runs both core and Travertino, while `-m test-core` and `-m test-trav` run them separately.

freakboy3742 reviewed (approved):
I've made a couple of minor tweaks, including promoting the release note to FEATURE, since it's a major change worth note - but I've also included a TODO in that feature which will act as a reminder when I draft the final Toga 0.5.0 release notes.

Otherwise this looks great! Thanks for all the work on this (and all the Travertino fixes that have made this change something worth doing!)

HalfWhitt commented:
And thank you for the cleanup tweaks! I've already got ideas bubbling for more things to improve in Travertino, and it'll be *so much easier* to do now. 😄
